File: checks/ruby/lib/section508/checks/placeholder_not_label.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Placeholder must not be the only label.
    class PlaceholderNotLabel
      SKIP_INPUT_TYPES = %w[hidden submit button reset image].freeze

      def self.call(root)
        return [] unless root

        offenders = placeholder_only_controls(root)
        return [] if offenders.empty?

        sample = Dom.field_hint(offenders.first)
        ["“#{sample}” uses placeholder as the only label."]
      end

      def self.placeholder_only_controls(root)
        return [] unless root.respond_to?(:css)

        Dom.elements(root, "input, select, textarea").select do |element|
          if element.name.upcase == "INPUT"
            type = (element["type"] || "text").downcase
            next false if SKIP_INPUT_TYPES.include?(type)
          end
          placeholder = element["placeholder"].to_s.strip
          next false if placeholder.empty?

          !Dom.has_associated_visible_label?(element, root)
        end
      end
    end
  end
end

File: checks/python/src/section508/checks/placeholder_not_label.py
from section508.checks.visible_label import (
    SKIP_INPUT_TYPES,
    field_hint,
    has_associated_visible_label,
)
from section508.dom import attr, select, tag_name


def check(root):
    if not root:
        return []

    offenders = placeholder_only_controls(root)
    if not offenders:
        return []

    sample = field_hint(offenders[0])
    return [f"“{sample}” uses placeholder as the only label."]


def placeholder_only_controls(root):
    found = []
    for element in select(root, "input, select, textarea"):
        if tag_name(element) == "input":
            type_name = (attr(element, "type") or "text").lower()
            if type_name in SKIP_INPUT_TYPES:
                continue
        placeholder = (attr(element, "placeholder") or "").strip()
        if not placeholder:
            continue
        if not has_associated_visible_label(element, root):
            found.append(element)
    return found
